src/core/public/application/application_service.tsx

```diff
@@ -311,6 +318,30 @@ export class ApplicationService {
         );
       },
     };
+
+    // do not add the global handler in legacy mode as it would perform a full refresh anyway
+    // also, there is a trick with 'multiple' legacy apps such as kibana:XXX we don't want to
+    // handle to avoid adding complexity to the handler's logic.
+    if (!this.legacyMode) {
+      this.linkClickDelegateHandler = e => {
+        const link = e.currentTarget as HTMLAnchorElement;
+        if (
+          link.href && // ignore links with empty hrefs
```

Comment:
should we also check for:

1. if `defaultPrevented` -> I guess we should skip navigation in this case?
2. if `link` has any `target` attr except `self` then we also should skip and allow browser to handle it?

Reply:
1. I'm unsure about that one. But as we are kinda 'replacing' the default behavior of the `a` element here, It probably makes sense. This is what is done in the `navlink` links anyway. @joshdover wdyt?

2. I think you are right.

Reply:
Maybe checking for `defaultPrevented` here removes the need for `'disableCoreNavigation'` ?

Reply:
When testing with the navlinks, it seems that our handler is executed before the event directly bound on the element with `onClick` handler. So unfortunately probably not.
Also the `disableCoreNavigation` allow to disable the behavior for a whole DOM tree. I agree it's so 2000's, but we have to keep in mind that core is framework agnostic and that  react is not the only way to go.

Reply:
Auch, but seems like our handler should be executed the last? Maybe it handle event on capturing phase instead of bubbling? 
If it is not executed last, then there is no way to prevent default inside `<a onClick` ?

Reply:
It should be executed last yea. We do want to be able to prevent this from a 'deeper' handler.

It has the option to use the capture phase, but it's during bubbling by default:

https://github.com/dgraham/delegated-events/blob/master/delegated-events.js#L103-L110

It was late yesterday when I checked. I will retest that today. I agree, not being able to prevent that from JS code would be problematic, but seeing the implem, nothing hacky/ugly is made, so we should be good.

Will add a FTR test for that also.

Reply:
Didn't look into implementation of the lib, but just wondering if the lib works like this for such case:

```
<a href="...">
  <span>I'm the content of the link</span>
<a/>
```

if user clicks `span` maybe the lib "synthetically" delegates the event all way to the top, so our global handler fires BEFORE native event bubbling starts 🤔 And we miss the onClick on link because of this.

Reply:
So, actually it seems to be caused by the way react bind its handlers, as I'm getting the same issue with a classic `document.addEventListener` call...

with 
```ts
    on('click', 'a', () => {
      console.log('`on` click handler');
    });
    document.addEventListener('click', () => {
      console.log('manual click handler');
    });
```

and 

```ts
      <a
        href={`/app/${targetAppId}#bang`}
        onClick={e => {
          console.log('direct handler');
        }}
      >
```

I'm getting, in order
```
`on` click handler
manual click handler
direct handler
```

So it  **seems**  we would not be able to cancel this behavior from a direct `onClick` handler as our handler will be executed first... I guess the `disableCoreNavigation` css class would be our only option here.

Not sure we should consider that as blocking or not to implement this solution... @joshdover WDYT?

Reply:
https://stackoverflow.com/questions/25862475/react-and-jquery-event-handlers-fired-in-wrong-order seems to be confirming this

> There are known issues with using jquery AND react events together. You'll find that often stopping propagation doesn't matter to either because the event isn't actually bound in javascript to the dom object you think it is, it's actually bound to the document and then handled from there behind the scenes

Meaning that react is also using a global `document.addEventListener` for all its events so we can't be sure of the actual executing order...

Reply:
> So it **seems** we would not be able to cancel this behavior from a direct `onClick` handler bound **from react** as our handler will be executed first... I guess the `disableCoreNavigation` css class would be our only option here.

I don't have any better ideas than using some attribute on the DOM element to opt-out. Would it make more sense to have a custom attribute like `data-kibana-disable-core-navigation` rather than a CSS class?

Reply:
An attribute seems cleaner and more explicit than a css class.